onnxruntime/core/optimizer/initializer.cc

```diff
@@ -11,49 +10,313 @@
 #include "core/framework/tensor_external_data_info.h"
 #include "core/platform/env.h"
 
+#include <functional>
+
 namespace onnxruntime {
-Status Initializer::ReadExternalRawData(
-    const ONNX_NAMESPACE::TensorProto& tensor_proto, const Path& model_path, std::vector<char>& raw_data) {
-  ORT_RETURN_IF_NOT(
-      tensor_proto.data_type() != ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED &&
-          tensor_proto.data_type() != ONNX_NAMESPACE::TensorProto_DataType_STRING,
-      "External data type must not be UNDEFINED or STRING.");
 
-  ORT_RETURN_IF(
-      model_path.IsEmpty(),
-      "model_path must not be empty. Ensure that a path is provided when the model is created or loaded.");
+Initializer::Initializer(ONNX_NAMESPACE::TensorProto_DataType data_type,
+                         std::string_view name,
+                         gsl::span<const int64_t> dims)
+    : name_(name),
+      data_([&]() {
+        Tensor w(DataTypeImpl::TensorTypeFromONNXEnum(data_type)->GetElementType(), dims, std::make_shared<CPUAllocator>());
+        if (!w.IsDataTypeString()) {
+          memset(w.MutableDataRaw(), 0, w.SizeInBytes());
+        }
+        return w;
+      }()) {
+}
+
+Initializer::Initializer(const ONNX_NAMESPACE::TensorProto& tensor_proto, const Path& model_path) {
+  ORT_ENFORCE(utils::HasDataType(tensor_proto), "Initializer must have a datatype");
+  if (utils::HasExternalData(tensor_proto)) {
+    ORT_ENFORCE(!model_path.IsEmpty(),
+                "model_path must not be empty. Ensure that a path is provided when the model is created or loaded.");
+  }
+
+  auto proto_data_type = tensor_proto.data_type();
+  if (utils::HasName(tensor_proto)) {
+    name_ = tensor_proto.name();
+  }
+
+  auto proto_dims = utils::GetTensorShapeFromTensorProto(tensor_proto);
+  TensorShape proto_shape(proto_dims);
+
+  // This must be pre-allocated
+  Tensor w(DataTypeImpl::TensorTypeFromONNXEnum(proto_data_type)->GetElementType(), proto_shape, std::make_shared<CPUAllocator>());
+  ORT_THROW_IF_ERROR(utils::TensorProtoToTensor(Env::Default(), model_path.ToPathString().c_str(), tensor_proto, w));
+  data_ = std::move(w);
+}
+
+namespace {
+template <typename T>
+struct ToFp16;
+
+template <>
+struct ToFp16<MLFloat16> {
+  uint16_t operator()(const MLFloat16& fl) const {
+    return fl.val;
+  }
+};
 
-  std::unique_ptr<ExternalDataInfo> external_data{};
-  ORT_RETURN_IF_ERROR(ExternalDataInfo::Create(tensor_proto.external_data(), external_data));
+template <>
+struct ToFp16<float> {
+  uint16_t operator()(float f) const {
+    return MLFloat16(f).val;
+  }
+};
 
-  size_t actual_tensor_data_length;
-  ORT_RETURN_IF_ERROR(utils::GetSizeInBytesFromTensorProto<0>(
-      tensor_proto, &actual_tensor_data_length));
-  const size_t external_data_length = external_data->GetLength();
+template <>
+struct ToFp16<double> {
+  uint16_t operator()(double d) const {
+    return MLFloat16(static_cast<float>(d)).val;
```

Comment:
Add comment why gsl::narrow isn't used/needed here?

Reply:
Because it will fail. We assume there will be a loss of precision, the same code as in Eigen.